File: php/game.php
<?php

class Game {
  // The game board and the game status
  public static $board = array("0", "1", "2", "3", "4", "5", "6", "7", "8");
  public static $currentState;  // the current state of the game
  public static $currentPlayer; // the current player

  /** The entry main method (the program starts here) */
  public static function main() {
    // Initialize the game-board and current status
    self::initGame();
    self::printBoard();
    do {
      self::getHumanSpot();
      if (!self::gameIsOver() && !self::tie()) {
        self::evalBoard();
      }
    } while (!self::gameIsOver() && !self::tie()); // repeat if not game-over
    echo("Game over\n");
  }

  /** Initializes the game */
  public static function initGame() {
    self::$currentState = 0; // "playing" or ready to play
    self::$currentPlayer = "X";  // cross plays first
  }

  public static function nextPlayer() {
    if (self::$currentPlayer == "X") {
      return "O";
    } else {
      return "X";
    }
  }

  /** Update global variables "board" and "currentPlayer". */
  public static function getHumanSpot() {
    $validInput = false;  // for input validation
    echo("Enter [0-8]:\n");
    do {
      $spot = stream_get_line(STDIN, 1024, PHP_EOL);
      if (self::$board[$spot] != "X" && self::$board[$spot] != "O") {
        self::$board[$spot] = "X";  // update game-board content
        self::printBoard();
        $validInput = true;  // input okay, exit loop
      }
      self::$currentPlayer = self::nextPlayer();  // cross plays first
    } while (!$validInput);  // repeat until input is valid
  }

  public static function evalBoard() {
    $foundSpot = false;
    do {
      if (self::$board[4] == "4") {
        self::$board[4] = "O";
        $foundSpot = true;
      } else {
        $spot = self::getBestMove();
        if (self::$board[$spot] != "X" && self::$board[$spot] != "O") {
          $foundSpot = true;
          self::$board[$spot] = "O";
        } else {
          $foundSpot = false;
        }
      }
    } while (!$foundSpot);
    self::printBoard();
  }

  /** Return true if the game was just won */
  public static function gameIsOver() {
    return self::$board[0] == self::$board[1] && self::$board[1] == self::$board[2] ||
      self::$board[3] == self::$board[4] && self::$board[4] == self::$board[5] ||
      self::$board[6] == self::$board[7] && self::$board[7] == self::$board[8] ||
      self::$board[0] == self::$board[3] && self::$board[3] == self::$board[6] ||
      self::$board[1] == self::$board[4] && self::$board[4] == self::$board[7] ||
      self::$board[2] == self::$board[5] && self::$board[5] == self::$board[8] ||
      self::$board[0] == self::$board[4] && self::$board[4] == self::$board[8] ||
      self::$board[2] == self::$board[4] && self::$board[4] == self::$board[6];
  }


  public static function getBestMove() {
    $availableSpaces = array();
    $foundBestMove = false;
    $spot = 100;
    foreach (self::$board as $s) {
      if ($s != "X" && $s != "O") {
        $availableSpaces[] = $s;
      }
    }
    foreach ($availableSpaces as $as) {
      $spot = intval($as);
      self::$board[$spot] = "O";
      if (self::gameIsOver()) {
        $foundBestMove = true;
        self::$board[$spot] = $as;
        return $spot;
      } else {
        self::$board[$spot] = "X";
        if (self::gameIsOver()) {
          $foundBestMove = true;
          self::$board[$spot] = $as;
          return $spot;
        } else {
          self::$board[$spot] = $as;
        }
      }
    }
    if ($foundBestMove) {
      return $spot;
    } else {
      $n = rand(0, count($availableSpaces));
      return intval($availableSpaces[$n]);
    }
  }

  /** Return true if it is a draw (no more empty cell) */
  // TODO: maybe there is an easeir way to check this
  public static function tie() {
    return self::$board[0] != "0" &&
           self::$board[1] != "1" &&
           self::$board[2] != "2" &&
           self::$board[3] != "3" &&
           self::$board[4] != "4" &&
           self::$board[5] != "5" &&
           self::$board[6] != "6" &&
           self::$board[7] != "7" &&
           self::$board[8] != "8";
  }

  /** Print the game board */
  public static function printBoard() {
    echo(" " . self::$board[0] . " | " . self::$board[1] . " | " . self::$board[2] . "\n===.===.===\n" . " " . self::$board[3] . " | " . self::$board[4] . " | " . self::$board[5] . "\n===.===.===\n" . " " . self::$board[6] . " | " . self::$board[7] . " | " . self::$board[8] . "\n\n"); // print all the self::$board cells
  }

}

Game::main();

File: java/Game.java
import java.util.Scanner;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Tic-Tac-Toe: TWo-player console version.
 */
public class Game {
  // The game board and the game status
  public static String[] board = {"0", "1", "2", "3", "4", "5", "6", "7", "8"};
  public static int currentState;  // the current state of the game
  public static String currentPlayer; // the current player

  public static Scanner input = new Scanner(System.in); // the input Scanner

  /** The entry main method (the program starts here) */
  public static void main(String[] args) {
    // Initialize the game-board and current status
    initGame();
    printBoard();
    do {
      getHumanSpot();
      if (!gameIsOver() && !tie()) {
        evalBoard();
      }
    } while (!gameIsOver() && !tie()); // repeat if not game-over
    System.out.print("Game over\n");
  }

  /** Initializes the game */
  public static void initGame() {
    currentState = 0; // "playing" or ready to play
    currentPlayer = "X";  // cross plays first
  }

  public static String nextPlayer() {
    if (currentPlayer == "X") {
      return "O";
    } else {
      return "X";
    }
  }

  /** Update global variables "board" and "currentPlayer". */
  public static void getHumanSpot() {
    boolean validInput = false;  // for input validation
    System.out.print("Enter [0-8]:\n");
    do {
      int spot = input.nextInt();
      if (board[spot] != "X" && board[spot] != "O") {
        board[spot] = "X";  // update game-board content
        printBoard();
        validInput = true;  // input okay, exit loop
      }
      currentPlayer = nextPlayer();  // cross plays first
    } while (!validInput);  // repeat until input is valid
  }

  public static void evalBoard() {
    boolean foundSpot = false;
    do {
      if (board[4] == "4") {
        board[4] = "O";
        foundSpot = true;
      } else {
        int spot = getBestMove();
        if (board[spot] != "X" && board[spot] != "O") {
          foundSpot = true;
          board[spot] = "O";
        } else {
          foundSpot = false;
        }
      }
    } while (!foundSpot);
    printBoard();
  }

  /** Return true if the game was just won */
  public static boolean gameIsOver() {
    return board[0] == board[1] && board[1] == board[2] ||
      board[3] == board[4] && board[4] == board[5] ||
      board[6] == board[7] && board[7] == board[8] ||
      board[0] == board[3] && board[3] == board[6] ||
      board[1] == board[4] && board[4] == board[7] ||
      board[2] == board[5] && board[5] == board[8] ||
      board[0] == board[4] && board[4] == board[8] ||
      board[2] == board[4] && board[4] == board[6];
  }


  public static int getBestMove() {
    ArrayList<String> availableSpaces = new ArrayList<String>();
    boolean foundBestMove = false;
    int spot = 100;
    for (String s: board) {
      if (s != "X" && s != "O") {
        availableSpaces.add(s);
      }
    }
    for (String as: availableSpaces) {
      spot = Integer.parseInt(as);
      board[spot] = "O";
      if (gameIsOver()) {
        foundBestMove = true;
        board[spot] = as;
        return spot;
      } else {
        board[spot] = "X";
        if (gameIsOver()) {
          foundBestMove = true;
          board[spot] = as;
          return spot;
        } else {
          board[spot] = as;
        }
      }
    }
    if (foundBestMove) {
      return spot;
    } else {
      int n = ThreadLocalRandom.current().nextInt(0, availableSpaces.size());
      return Integer.parseInt(availableSpaces.get(n));
    }
  }

  /** Return true if it is a draw (no more empty cell) */
  // TODO: maybe there is an easeir way to check this
  public static boolean tie() {
    return board[0] != "0" &&
           board[1] != "1" &&
           board[2] != "2" &&
           board[3] != "3" &&
           board[4] != "4" &&
           board[5] != "5" &&
           board[6] != "6" &&
           board[7] != "7" &&
           board[8] != "8";
  }

  /** Print the game board */
  public static void printBoard() {
    System.out.println(" " + board[0] + " | " + board[1] + " | " + board[2] + "\n===+===+===\n" + " " + board[3] + " | " + board[4] + " | " + board[5] + "\n===+===+===\n" + " " + board[6] + " | " + board[7] + " | " + board[8] + "\n"); // print all the board cells
  }

}
